### PR TITLE
Omni shadows for clustered lights

### PR DESCRIPTION
**Pull request by mvaligursky**
Clustered omni lights now support shadows by rendering them to the shadow atlas (shared with the spot lights)
- webgl1 encodes depth in color buffer and does manual PCF comparison
- webgl2 uses depth shadows sampler and writes custom depth to the shadow map, and uses HW PCF
- it uses PCF3x3 (the same as spot lights), but single sample version is implemented as well and will be exposed at a later stage.
- omni light uses few pixel larger FOV than 90deg to avoid doing PCF filtering across multiple faces.
- new example to demonstrate: clustered-omni-shadows.tsx

Screenshot 2021-10-05 at 12 15 47

### CODE REVIEW

**Comment by Maksims** [comment]
This is amazing!

I've checked recently clustered for Omni lights and noticed it supports only linear falloff.
Also, I assume specularity is a bit more complicated, as it can extend outside of the light range?

**Comment by mvaligursky** [comment]
Correct @Maksims , the lighting model is very limited at the moment, just a basic diffuse. This is what's still to do:
https://github.com/playcanvas/engine/issues/3522

File: src/graphics/program-lib/chunks/cookie.frag [hunk 1]
```diff
@@ -37,5 +37,5 @@ vec3 getCookie2DClustered(sampler2D tex, mat4 transform, float intensity, bool i
     vec4 projPos = transform * vec4(vPositionW, 1.0);
     projPos.xy /= projPos.w;
     vec4 pixel = mix(vec4(1.0), texture2D(tex, projPos.xy), intensity);
-    return isRgb ? pixel.rgb : vec3(dot(pixel, cookieChannel));
+    return isRgb == true ? pixel.rgb : vec3(dot(pixel, cookieChannel));
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/cookie.frag:40]: Just wondering why this change is needed.

Comment (willeastcott) [inline src/graphics/program-lib/chunks/cookie.frag:40]: Why do you need to add the `== true` here?

File: src/graphics/program-lib/chunks/shadowStandard.frag [hunk 2]
```diff
@@ -180,3 +180,116 @@ float _getShadowPoint(samplerCube shadowMap, vec4 shadowParams, vec3 dir) {
 float getShadowPointPCF3x3(samplerCube shadowMap, vec4 shadowParams) {
     return _getShadowPoint(shadowMap, shadowParams, dLightDirW);
 }
+
+// ----- Clustered Omni Sampling using atlas -----
+
+// Converts unnormaized direction vector to a cubemap face index [0..5] and uv coordinates within the face in [0..1] range.
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowStandard.frag:186]:
```suggestion
// Converts unnormalized direction vector to a cubemap face index [0..5] and uv coordinates within the face in [0..1] range.
```

File: src/graphics/program-lib/chunks/shadowStandard.frag [hunk 3]
```diff
@@ -180,3 +180,116 @@ float _getShadowPoint(samplerCube shadowMap, vec4 shadowParams, vec3 dir) {
...
+
+		faceIndex = dir.x < 0.0 ? 1.0 : 0.0;
+		ma = 0.5 / vAbs.x;
+		uv = vec2(dir.x < 0.0 ? dir.z : -dir.z, -dir.y);
+
+        tileOffset.x = 0.0;
+        tileOffset.y = dir.x < 0.0 ? 1.0 : 0.0;
+
+	}
+	return uv * ma + 0.5;
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowStandard.frag:221]: Looks like a mixture of tabs and spaces have crept in here. I guess we've standardized on spaces?

File: src/graphics/program-lib/chunks/shadowStandard.frag [hunk 4]
```diff
@@ -180,3 +180,116 @@ float _getShadowPoint(samplerCube shadowMap, vec4 shadowParams, vec3 dir) {
...
+}
+
+// converts unnormaized direction vector to a texture coordinate for a cubemap face stored within texture atlas described by the viewport
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowStandard.frag:224]:
```suggestion
// converts unnormalized direction vector to a texture coordinate for a cubemap face stored within texture atlas described by the viewport
```

File: src/graphics/program-lib/programs/standard.js [hunk 5]
```diff
@@ -477,6 +477,92 @@ var standard = {
...
+                code += "   float minValue = 2.3374370500153186e-10; //(1.0 / 255.0) / (256.0 * 256.0 * 256.0);\n";
+                code += "   depth += polygonOffset.x * max(abs(dFdx(depth)), abs(dFdy(depth))) + minValue * polygonOffset.y;\n";
+                code += "   gl_FragColor = packFloat(depth);\n";
+            } else {
+                code += "   gl_FragColor = packFloat(depth);\n";
+            }
+        } else if (shadowType === SHADOW_PCF3 || shadowType === SHADOW_PCF5) {
+            code += "   gl_FragColor = vec4(1.0);\n"; // just the simpliest code, color is not written anyway
+
+            // clustered omni light is using shadow sampler and neesd to write custom depth
```
Comment (willeastcott) [inline src/graphics/program-lib/programs/standard.js:551]:
```suggestion
            // clustered omni light is using shadow sampler and needs to write custom depth
```

**Review by willeastcott — APPROVED** [review]

File: src/graphics/program-lib/chunks/cookie.frag (hunk 1 above)
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/cookie.frag:40]: to be honest, I'm not sure, but have noticed one branch was not taken correctly unless I added it. So I added it to all. I could not find a different workaround.

File: src/graphics/program-lib/chunks/shadowStandard.frag (hunk 3 above)
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/shadowStandard.frag]: done, thanks

File: src/graphics/program-lib/chunks/cookie.frag (hunk 1 above)
Comment (willeastcott) [reply src/graphics/program-lib/chunks/cookie.frag:40]: TBH, if I was refactoring this code, I would delete this. So if it's important, I would add a comment to say why it's there.
